quodlibet/order/__init__.py

```diff
@@ -103,39 +105,42 @@ class OrderRemembered(Order):
     """Shared class for all the shuffle modes that keep a memory
     of their previously played songs."""
 
+    _played: List[Gtk.TreeIter]
+
     def __init__(self):
         super().__init__()
         self._played = []
 
     def next(self, playlist, iter):
         if iter is not None:
-            self._played.append(playlist.get_path(iter).get_indices()[0])
+            self._played.append(iter)
 
     def previous(self, playlist, iter):
-        try:
-            path = self._played.pop()
-        except IndexError:
-            return None
-        else:
-            return playlist.get_iter(path)
+        if self._played:
+            return self._played.pop()
+        return None
 
     def set(self, playlist, iter):
         if iter is not None:
-            self._played.append(playlist.get_path(iter).get_indices()[0])
+            self._played.append(iter)
         return iter
 
     def reset(self, playlist):
         del(self._played[:])
 
-    def remaining(self, playlist):
+    def remaining(self, playlist) -> Dict[int, Any]:
         """Gets a map of all song indices to their song from the `playlist`
         that haven't yet been played"""
-        all_indices = set(range(len(playlist)))
-        played = set(self._played)
+
+        def get_index(iter):
+            return playlist.get_path(iter).get_indices()[0]
```

Comment:
I remember this line coming up a bit whilst debugging. Can we put some null-saftety (etc) here?

If somehow `playlists.get_path(iter)` returns `None`. Or `get_indices()` does, or is empty?...
If they're at least on separate lines, stack traces are a bit quicker to read

Reply:
I'll split up the expression into multiple lines. Other than that, I'm worried that not throwing on `None` or an empty list will make it harder to discover incorrect code.